sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/timewindow/recurrence/RecurrenceEvaluator.java

```diff
@@ -0,0 +1,158 @@
+        ZonedDateTime start = settings.getStart();
+        OccurrenceInfo emptyOccurrence = new OccurrenceInfo();
+        if (now.isBefore(start)) {
+            return emptyOccurrence;
+        }
+
+        final RecurrencePatternType patternType = settings.getRecurrence().getPattern().getType();
+        OccurrenceInfo occurrenceInfo;
+        if (patternType == RecurrencePatternType.DAILY) {
+            occurrenceInfo = getDailyPreviousOccurrence(settings, now);
+        } else {
+            occurrenceInfo = getWeeklyPreviousOccurrence(settings, now);
+        }
+
+        final RecurrenceRange range = settings.getRecurrence().getRange();
+        if (range.getType() == RecurrenceRangeType.ENDDATE
+            && occurrenceInfo.previousOccurrence != null
+            && occurrenceInfo.previousOccurrence.isAfter(range.getEndDate())) {
+            return emptyOccurrence;
+        }
+        if (range.getType() == RecurrenceRangeType.NUMBERED
+            && occurrenceInfo.numberOfOccurrences > range.getNumberOfRecurrences()) {
+            return emptyOccurrence;
+        }
+
+        return occurrenceInfo;
+    }
+
+    /**
+     * Find the closest previous recurrence occurrence before the provided time stamp according to the "Daily" recurrence pattern.
+     *
+     * @return The return result contains two property, one is previousOccurrence, the other is numberOfOccurrences.
+     * previousOccurrence: The closest previous occurrence.
+     * numberOfOccurrences: The number of complete recurrence intervals which have occurred between the time and the recurrence start.
+     */
+    private static OccurrenceInfo getDailyPreviousOccurrence(TimeWindowFilterSettings settings, ZonedDateTime now) {
+        final ZonedDateTime start = settings.getStart();
+        final int interval = settings.getRecurrence().getPattern().getInterval();
+        final int numberOfOccurrences = (int) (Duration.between(start, now).getSeconds() / Duration.ofDays(interval).getSeconds());
+        return new OccurrenceInfo(start.plusDays((long) numberOfOccurrences * interval), numberOfOccurrences + 1);
+    }
+
+    /**
+     * Find the closest previous recurrence occurrence before the provided time stamp according to the "Weekly" recurrence pattern.
+     *
+     * @return The return result contains two property, one is previousOccurrence, the other is numberOfOccurrences.
+     * previousOccurrence: The closest previous occurrence.
+     * numberOfOccurrences: The number of recurring days of week which have occurred between the time and the recurrence start.
+     */
+    private static OccurrenceInfo getWeeklyPreviousOccurrence(TimeWindowFilterSettings settings, ZonedDateTime now) {
+        final RecurrencePattern pattern = settings.getRecurrence().getPattern();
+        final int interval = pattern.getInterval();
+        final ZonedDateTime firstDayOfFirstWeek = settings.getStart().minusDays(
```

Comment:
Can we have a variable for settings.getStart() instead of calling it repeatedly?

Reply:
Updated